noisy_simulator/src/kernel.rs

```diff
@@ -54,22 +60,44 @@ pub fn apply_kernel(
     let mut new_entries = ComplexVector::zeros(num_elements);
     for s in 0..state.len() {
         if (s & mask) == 0 {
-            new_entries.fill(Complex::ZERO);
-
             // Extract relevant entries into a vector to make the gate application easier.
             for k in 0..num_elements {
```

Comment:
Need to double check, but if you were to use `index_offsets.len()` the compiler should be able to infer get_unchecked without resorting to explicitly calling unsafe code.

Reply:
Indeed, I was able to remove the first `get_unchecked(k)` while still skipping the bound check. Just doing `for k in 0..index_offsets.len()` didn't work. I had to do `(0..index_offsets).for_each(|k| { ... })` to be able to remove the bound check there.

I tried several approaches to try to remove the `extracted_entries.get_unchecked_mut(k)` on the second unsafe block, but nothing worked. Here is a list of the things I tried:
 - An assert `assert!(index_offsets.len() == extracted_entries.len())` at the beginning of the loop.
 - An assert at the beginning of the outer loop.
 - `for _ in _` zipped iteration.
 - `for_each` zipped iteration.